### PR TITLE
Add base URL support

### PR DESCRIPTION
**Pull request by hungngo97**
Added base URL support follow up to issue in editor: [#686](https://github.com/vega/editor/issues/686)

### CODE REVIEW

**Comment by hungngo97** [comment]
Currently the tests on Github Actions still say fail though the log did show a correct SVG output, probably Github Actions use a different font scheme? One quick fix is `jest --updateSnapshot` to work with Github Actions font. But currently all the Jest tests are passing on my end.

**Comment by domoritz** [comment]
Maybe install the same fonts on github actions.

File: src/app.ts [hunk 1]
```diff
@@ -46,7 +46,13 @@ app.get('/', (req: Request, res: Response) => {
...
 app.post('/', async (req: Request, res: Response) => {
   const contentType = req.header('Accept') ?? 'pdf';
-  let spec = req.body;
+  if (!req.body.spec) {
+    return res
+        .status(400)
+        .end('Must provide Vega spec for render servicce');
+  }
+  let { spec } = req.body;
+  let baseURL = req.body.baseURL || '';
```
Comment (domoritz) [inline src/app.ts:55]:
```suggestion
  let { spec, baseURL } = req.body;
```

File: src/app.ts [hunk 2]
```diff
@@ -72,15 +81,26 @@ app.post('/', async (req: Request, res: Response) => {
       res.status(400).send('External URI not allowed on this API');
       throw new Error('External data url not allowed');
```
Comment (domoritz) [inline src/app.ts:82]: Do we need both?

File: src/constants.ts [hunk 3]
```diff
@@ -1 +1,4 @@
-export const ALLOWED_URLS: string[] = ['vega.github.io'];
+export const ALLOWED_URLS: string[] = [
+    'vega.github.io',
+    'raw.githubusercontent.com' // For testing purposes
```
Comment (domoritz) [inline src/constants.ts:3]: Do we still need this?

**Comment by domoritz** [comment]
I'll mark this pull request as draft until you're done testing.

**Comment by hungngo97** [comment]
Yup sorry, there is still somehow a mismatch between the snapshots between my local computer and github tests

**Comment by domoritz** [comment]
I'm sure this is coming from fonts. You could try adding support for docker with https://code.visualstudio.com/docs/remote/containers. Then you could run on ubuntu and have the same font behavior.

**Comment by hungngo97** [comment]
Yea I would need to do that fonts eventually but the current problem is just that the final size of the svg is not matching with my local svg snapshot, the rest of the svg is still the same
image

**Comment by domoritz** [comment]
Vega measures fonts and uses it to determine the layout of marks and guides. So if you don't have a particular font, the measurements can be different.

**Comment by hungngo97** [comment]
Ohhhhh okay, so let me investigate on using docker and see if I can install default fonts. Thank you. I was trying to see why is the size not deterministic :)

**Comment by hungngo97** [comment]
@domoritz I think it is ready for review. Setting up Docker helps the Github actions Jest testing and also install fonts as well. I also added some more command for now deployment to have more fonts and it looks great to me.
You can try my test deployment [here] (https://vega-render-service.hungngovinh97.now.sh/)

File: src/app.ts [hunk 4]
```diff
@@ -46,7 +46,13 @@ app.get('/', (req: Request, res: Response) => {
 
 app.post('/', async (req: Request, res: Response) => {
   const contentType = req.header('Accept') ?? 'pdf';
-  let spec = req.body;
+  if (!req.body.spec) {
+    return res
+        .status(400)
+        .end('Must provide Vega spec for render servicce');
```
Comment (domoritz) [inline src/app.ts:52]:
```suggestion
        .end('Must provide Vega spec for render service');
```

File: package-lock.json [hunk 5]
```diff
@@ -0,0 +1,7714 @@
+{
```
Comment (domoritz) [inline package-lock.json:1]: Remove this file

File: src/app.ts [hunk 6]
```diff
@@ -72,15 +81,26 @@ app.post('/', async (req: Request, res: Response) => {
...
+    try {
+      if (options) {
+        return await originalLoad(url, {
+          ...options,
+          ...{ baseURL }
+        });
+      }
+      return await originalLoad(url, { baseURL });
+    } catch {
+      return await originalLoad(url, options);
```
Comment (domoritz) [inline src/app.ts:97]: What's the default base url? I think it should be `https://vega.github.io/vega-datasets/`.

File: test/index.test.ts [hunk 7]
```diff
@@ -4,7 +4,10 @@ import request from 'supertest';
 import app from '../src/app';
 export const vegaSpec = require('../vegaSpecs/bar.vg.json');
 export const vegaliteSpec = require('../vegaSpecs/bar.vl.json');
-export const specUseExternalLink = require('../vegaSpecs/specUseExternalLink.vg.json');
+export const specUseExternalLink = require('../vegaSpecs/specUseInvalidExternalLink.vg.json');
+export const vegaSpecWithRelativeUrl = require('../vegaSpecs/specWithRelativeUrl.vg.json');
+export const vegaliteSpecWithRelativeUrl = require('../vegaSpecs/specWithRelativeUrl.vl.json');
+export const VEGA_GITHUB_RAW_DATA_BASE_URL = 'https://raw.githubusercontent.com/vega/vega/master/docs/';
```
Comment (domoritz) [inline test/index.test.ts:10]: Why not `https://vega.github.io/vega-datasets/`?

File: package.json [hunk 8]
```diff
@@ -6,7 +6,7 @@
   "scripts": {
     "build": "tsc",
     "prestart": "npm run build",
-    "now-build": "yum install gcc-c++ cairo-devel pango-devel libjpeg-turbo-devel giflib-devel && npm run build",
+    "now-build": "yum install gcc-c++ cairo-devel pango-devel libjpeg-turbo-devel giflib-devel libuuid-devel libmount-devel && cp /lib64/{libuuid,libmount,libblkid}.so.1 node_modules/canvas/build/Release/ && npm run build",
```
Comment (domoritz) [inline package.json:9]: Why is this change needed?

File: src/app.ts (hunk 6 above)
Comment (hungngo97) [reply src/app.ts:97]: The default baseURL is sent from the editor front end so I didn't hard code it

Comment (hungngo97) [reply src/app.ts:97]: If the user did not provide any base URL, shouldn't it be empty?

Comment (domoritz) [reply src/app.ts:97]: No, I think if the user does not provide one, we should use `https://vega.github.io/vega-datasets/`. That's more consistent with the editor.

File: src/app.ts [hunk 9]
```diff
@@ -49,10 +49,10 @@ app.post('/', async (req: Request, res: Response) => {
   if (!req.body.spec) {
     return res
         .status(400)
-        .end('Must provide Vega spec for render servicce');
+        .end('Must provide Vega spec for render service');
   }
   let { spec } = req.body;
-  let baseURL = req.body.baseURL || '';
+  let baseURL = req.body.baseURL || 'https://vega.github.io/vega-datasets/';
```
Comment (domoritz) [inline src/app.ts:55]: Let's make this a constant that you also use in the tests.

File: package.json (hunk 8 above)
Comment (hungngo97) [reply package.json:9]: Yea somehow there was an error in Now deployment that didn't have `uuid` so I need to install those extra packages

Comment (domoritz) [reply package.json:9]: That doesn't look right to me. It wasn't required in the past, right?

Comment (hungngo97) [reply package.json:9]: Yea it wasn't. This is just a problem with Now deployment. I followed the open issue [here](https://github.com/vercel/vercel/issues/3460) and copy the command.

**Comment by domoritz** [comment]
Is this is ready, please mark it as "ready for review".

Screen Shot 2020-06-14 at 18 37 15

**Comment by hungngo97** [comment]
I just added it. Thanks!

File: Dockerfile.build [hunk 10]
```diff
@@ -0,0 +1,29 @@
...
+    python \
+    ;
+
+RUN npm install
+
+COPY . .
+
+EXPOSE 8090
+CMD [ "npm", "run", "build" ]
+CMD ["jest", "--updateSnapshot"]
```
Comment (domoritz) [inline Dockerfile.build:27]:
```suggestion
CMD [ "jest", "--updateSnapshot" ]
```

File: src/app.ts [hunk 11]
```diff
@@ -46,7 +46,11 @@ app.get('/', (req: Request, res: Response) => {
 
 app.post('/', async (req: Request, res: Response) => {
   const contentType = req.header('Accept') ?? 'pdf';
-  let spec = req.body;
+  if (!req.body.spec) {
+    return res.status(400).end('Must provide Vega spec for render service');
+  }
+  let { spec } = req.body;
+  const baseURL = req.body.baseURL || VEGA_DATA_BASE_URL;
```
Comment (domoritz) [inline src/app.ts:53]:
```suggestion
  const baseURL = req.body.baseURL ?? VEGA_DATA_BASE_URL;
```
